### PR TITLE
Add onDidChange() method

### PR DESCRIPTION
**Pull request by callodacity**
This is a new PR to fix the mess I made in #22 -- sorry about that!

I did make some changes, which we can discuss:

- Added type checking to arguments
- Removed support for listening to changes without keys
- ~~Added `key` as a third argument to the callback~~

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -70,6 +73,30 @@ class Conf {
...
+
+		const onChange = () => {
+			const oldValue = currentValue;
+			const newValue = key ? this.get(key) : this.store;
+
+			try {
+				assert.deepEqual(newValue, oldValue);
+			} catch (err) {
+				currentValue = newValue;
+				callback.call(this, newValue, oldValue, key);
```
Comment (kevva) [inline index.js]: It doesn't make sense to have `key` as the third argument when the user already knows about it, IMO.

File: index.js [hunk 2]
```diff
@@ -70,6 +73,30 @@ class Conf {
 	clear() {
 		this.store = obj();
 	}
+	onDidChange(key, callback) {
+		if (typeof key !== 'string') {
+			throw new TypeError(`Expected \`key\` to be of type \`string\`, got ${typeof key}`);
+		} else if (typeof callback !== 'function') {
```
Comment (kevva) [inline index.js]: Don't need `else if` here. Just do a new `if` block.

File: index.js [hunk 3]
```diff
@@ -70,6 +73,30 @@ class Conf {
...
+			throw new TypeError(`Expected \`callback\` to be of type \`function\`, got ${typeof callback}`);
+		}
+
+		let currentValue = key ? this.get(key) : this.store;
```
Comment (kevva) [inline index.js]: I don't think it should fall back to `this.store`. Just do `let currentValue = this.get(key);` and the same with `newValue` further down.

**Comment by SamVerschueren** [comment]
This is a little off topic and is more a discussion on the API of this feature, not really regarding the PR. Did we consider using Observables for this use case? 

```js
const Conf = require('conf');

const config = new Conf();

config.select('unicorn').subscribe(value => {
    console.log(value);
    //=> 🌈
});

config.set('unicorn', '🌈');
```

The benefit is that you can easily integrate it together with other observables or use other operators to manipulate the result before returning.

The `select` method is based upon how [ngrx](https://github.com/ngrx/platform) (Redux for Angular) does it. We can always use another name off course.

It's just an idea though. Want to make sure we considered all our options :).

**Comment by SamVerschueren** [comment]
Another approach could be like we did with [cache-conf](https://github.com/samverschueren/cache-conf). We keep `conf` small and focused, and this behaviour could be put in an `observable-conf` package which is a drop-in replacement for `conf`.

**Comment by SamVerschueren** [comment]
I discussed this with @sindresorhus and decided that this feature should be in the core. However, it's nice if we have a wrapper for observables, so I created [observable-conf](https://github.com/SamVerschueren/observable-conf). I can use this `onDidChange()` feature when it lands.

**Comment by callodacity** [comment]
@SamVerschueren no worries, that sounds good to me. Do you know if we are we waiting for anything before merging this into core?

**Comment by sindresorhus** [comment]
@callodacity Was just waiting on me having time to merge it. Thanks for working on this 😀

**Pull request merged**